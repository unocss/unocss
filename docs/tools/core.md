---
title: Core
description: The core engine of UnoCSS without any presets. It can be used as the engine of your own atomic CSS framework.
---

# Core

The core engine of UnoCSS without any presets: `@unocss/core`. It can be used as the engine of your own atomic CSS framework.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D @unocss/core
```

```bash [yarn]
yarn add -D @unocss/core
```

```bash [npm]
npm install -D @unocss/core
```

```bash [bun]
bun add -D @unocss/core
```

:::

## Usage

```ts
import { createGenerator } from '@unocss/core'

const generator = await createGenerator(
  { /* user options */ },
  { /* default options */ }
)

const { css } = await generator.generate(code)
```

## License

- MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
