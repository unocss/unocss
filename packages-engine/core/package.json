{
  "name": "@unocss/core",
  "type": "module",
  "version": "66.1.0-beta.3",
  "description": "The instant on-demand Atomic CSS engine.",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://unocss.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss",
    "directory": "packages-engine/core"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub",
    "test:attw": "attw --pack --config-path ../../.attw-esm-only.json"
  },
  "devDependencies": {
    "magic-string": "catalog:",
    "unconfig": "catalog:"
  }
}
