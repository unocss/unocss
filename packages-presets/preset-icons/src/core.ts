import type { IconifyJSON } from '@iconify/types'
import type { IconifyLoaderOptions, UniversalIconLoader } from '@iconify/utils'
import type { CSSObject } from '@unocss/core'
import type { IconsOptions } from './types'
import { loadIcon } from '@iconify/utils/lib/loader/loader'
import { searchForIcon } from '@iconify/utils/lib/loader/modern'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'
import { definePreset, warnOnce } from '@unocss/core'
import icons from './collections'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }
export { icons }

/**
 * API for preset-icons
 */
export interface IconsAPI {
  encodeSvgForCss: typeof encodeSvgForCss
  parseIconWithLoader: typeof parseIconWithLoader
  /**
   * This API only available for preset-icons created on Node.js environment
   */
  createNodeLoader?: () => Promise<UniversalIconLoader | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return definePreset((options: IconsOptions = {}) => {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      iconifyCollectionsNames,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      collectionsNodeResolvePath,
      layer = 'icons',
      unit,
      processor,
    } = options

    const flags = getEnvFlags()

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      cwd: collectionsNodeResolvePath,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    let iconLoader: UniversalIconLoader

    return {
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      api: <IconsAPI>{
        encodeSvgForCss,
        parseIconWithLoader,
      },
      rules: [[
        /^([\w:-]+)(?:\?(mask|bg|auto))?$/,
        async (matcher) => {
          let [full, body, _mode = mode] = matcher as [string, string, IconsOptions['mode']]

          iconLoader = iconLoader || await lookupIconLoader(options)

          const usedProps = {}
          const parsed = await parseIconWithLoader(
            body,
            iconLoader,
            { ...loaderOptions, usedProps },
            iconifyCollectionsNames,
          )

          if (!parsed) {
            if (warn && !flags.isESLint)
              warnOnce(`failed to load icon "${full}"`)
            return
          }

          let cssObject: CSSObject
          const url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(parsed.svg)}")`

          if (_mode === 'auto')
            _mode = parsed.svg.includes('currentColor') ? 'mask' : 'bg'

          if (_mode === 'mask') {
            // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
            cssObject = {
              '--un-icon': url,
              '-webkit-mask': 'var(--un-icon) no-repeat',
              'mask': 'var(--un-icon) no-repeat',
              '-webkit-mask-size': '100% 100%',
              'mask-size': '100% 100%',
              'background-color': 'currentColor',
              // for Safari https://github.com/elk-zone/elk/pull/264
              'color': 'inherit',
              ...usedProps,
            }
          }
          else {
            cssObject = {
              'background': `${url} no-repeat`,
              'background-size': '100% 100%',
              'background-color': 'transparent',
              ...usedProps,
            }
          }

          processor?.(cssObject, {
            ...parsed,
            icon: parsed.name,
            mode: _mode,
          })

          return cssObject
        },
        { layer, prefix },
      ]],
    }
  })
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

export function createCDNFetchLoader(
  fetcher: (url: string) => Promise<any>,
  cdnBase: string,
  cacheMap: Map<string, IconifyJSON | Promise<IconifyJSON>> = new Map(),
): UniversalIconLoader {
  function fetchCollection(name: string) {
    if (!icons.includes(name))
      return undefined
    if (!cacheMap.has(name))
      cacheMap.set(name, fetcher(`${cdnBase}@iconify-json/${name}/icons.json`))
    return cacheMap.get(name)!
  }

  return async (collection, icon, options) => {
    let result = await loadIcon(collection, icon, options)
    if (result)
      return result

    const iconSet = await fetchCollection(collection)
    if (iconSet) {
      // possible icon names
      const ids = [
        icon,
        icon.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
        icon.replace(/([a-z])(\d+)/g, '$1-$2'),
      ]
      result = await searchForIcon(iconSet, collection, ids, options)
    }

    return result
  }
}

export function getEnvFlags() {
  // eslint-disable-next-line node/prefer-global/process
  const isNode = typeof process !== 'undefined' && process.stdout && !process.versions.deno
  // eslint-disable-next-line node/prefer-global/process
  const isVSCode = isNode && !!process.env.VSCODE_CWD
  // eslint-disable-next-line node/prefer-global/process
  const isESLint = isNode && !!process.env.ESLINT

  return {
    isNode,
    isVSCode,
    isESLint,
  }
}

export async function parseIconWithLoader(
  body: string,
  loader: UniversalIconLoader,
  options: IconifyLoaderOptions = {},
  safeCollectionsNames: string[] = [],
) {
  let collection = ''
  let name = ''
  let svg: string | undefined

  const allCollections = new Set<string>([
    ...icons,
    ...safeCollectionsNames,
    ...Object.keys(options.customCollections || {}),
  ])

  if (body.includes(':')) {
    [collection, name] = body.split(':')

    if (!allCollections.has(collection))
      return

    svg = await loader(collection, name, options)
  }
  else {
    const parts = body.split(/-/g)
    for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
      collection = parts.slice(0, i).join('-')

      if (!allCollections.has(collection))
        continue

      name = parts.slice(i).join('-')
      svg = await loader(collection, name, options)

      if (svg)
        break
    }
  }

  if (!svg) {
    return
  }

  return {
    collection,
    name,
    svg,
  }
}
