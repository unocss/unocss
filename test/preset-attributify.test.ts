import { createGenerator, toEscapedSelector as e } from '@unocss/core'
import { autocompleteExtractorAttributify, presetAttributify, variantAttributify } from '@unocss/preset-attributify'
import presetWind3 from '@unocss/preset-wind3'
import { describe, expect, it } from 'vitest'

describe('attributify', async () => {
  const uno = await createGenerator({
    presets: [
      presetAttributify({ strict: true }),
      presetWind3({ attributifyPseudo: true }),
    ],
    rules: [
      [/^custom-(\d+)$/, ([_, value], { rawSelector }) => {
        // return a string instead of an object
        const selector = e(rawSelector)
        return `
  ${selector} {
    font-size: ${value}px;
  }
  /* you can have multiple rules */
  ${selector}::after {
    content: 'after';
  }
  .foo > ${selector} {
    color: red;
  }
  /* or media queries */
  @media (min-width: 680px) {
    ${selector} {
      font-size: 16px;
    }
  }
  `
      }],
    ],
  })

  describe('cases', () => {
    const cases = import.meta.glob<string>('./cases/preset-attributify/*/input.html', { query: '?raw', import: 'default' })
    for (const [path, input] of Object.entries(cases)) {
      it(path, async () => {
        const { css, matched } = await uno.generate(await input(), { preflights: false })
        await expect(`${[...matched].join('\n')}\n`)
          .toMatchFileSnapshot(path.replace('input.html', 'matched.txt'))
        await expect(css)
          .toMatchFileSnapshot(path.replace('input.html', 'output.css'))
      })
    }
  })

  const fixture1 = (await import('./cases/preset-attributify/case-1/input.html?raw')).default

  it('variant', async () => {
    const variant = variantAttributify({
      prefix: 'un-',
      prefixedOnly: false,
      strict: true,
    })

    const promises = Array.from(await uno.applyExtractors(fixture1) || [])
      .map(async i => await variant.match(i, {} as any))

    expect((await Promise.all(promises))
      .flat()
      .map(r => typeof r === 'string' ? r : r ? r.matcher : r),
    )
      .toMatchSnapshot()
  })

  it('prefixedOnly', async () => {
    const uno = await createGenerator({
      presets: [
        presetAttributify({ strict: true, prefix: 'un-', prefixedOnly: true }),
        presetWind3({ attributifyPseudo: true }),
      ],
    })

    const { css } = await uno.generate(fixture1, { preflights: false })
    expect(css).toMatchSnapshot()
  })

  describe('autocomplete extractor', async () => {
    it('without prefix', async () => {
      const res = await autocompleteExtractorAttributify().extract({
        content: fixture1,
        cursor: 187,
      })

      expect(res).not.toBeNull()

      expect(res!.extracted).toMatchInlineSnapshot('"bg-blue-400"')
      expect(res!.transformSuggestions!([`${res!.extracted}1`, `${res!.extracted}2`]))
        .toMatchInlineSnapshot(`
          [
            "blue-4001",
            "blue-4002",
          ]
        `)

      const reversed = res!.resolveReplacement(`${res!.extracted}1`)
      expect(reversed).toMatchInlineSnapshot(`
        {
          "end": 192,
          "replacement": "blue-4001",
          "start": 184,
        }
      `)

      expect(fixture1.slice(reversed.start, reversed.end))
        .toMatchInlineSnapshot('"blue-400"')
    })

    it('with prefix', async () => {
      const fixtureWithPrefix = `
<div un-text-cent>
  <div un-text="cent
</div>
      `
      const res1 = await autocompleteExtractorAttributify({ prefix: 'un-' }).extract({
        content: fixtureWithPrefix,
        cursor: 18,
      })

      expect(res1).not.toBeNull()

      expect(res1!.extracted).toMatchInlineSnapshot('"text-cent"')

      const reversed1 = res1!.resolveReplacement(`${res1!.extracted}1`)
      expect(reversed1).toMatchInlineSnapshot(`
        {
          "end": 18,
          "replacement": "text-cent1",
          "start": 9,
        }
      `)

      expect(fixtureWithPrefix.slice(reversed1.start, reversed1.end))
        .toMatchInlineSnapshot('"text-cent"')

      const res2 = await autocompleteExtractorAttributify({ prefix: 'un-' }).extract({
        content: fixtureWithPrefix,
        cursor: 40,
      })

      expect(res2).not.toBeNull()

      expect(res2!.extracted).toMatchInlineSnapshot('"text-cent"')
      expect(res2!.transformSuggestions!([`${res2!.extracted}1`, `${res2!.extracted}2`]))
        .toMatchInlineSnapshot(`
          [
            "cent1",
            "cent2",
          ]
        `)

      const reversed2 = res2!.resolveReplacement(`${res2!.extracted}1`)
      expect(reversed2).toMatchInlineSnapshot(`
        {
          "end": 40,
          "replacement": "cent1",
          "start": 36,
        }
      `)

      expect(fixtureWithPrefix.slice(reversed2.start, reversed2.end))
        .toMatchInlineSnapshot('"cent"')
    })

    it('only prefix', async () => {
      const fixtureOnlyPrefix = `
<div text-cent>
  <div text="cent
</div>
      `
      const res1 = await autocompleteExtractorAttributify({ prefix: 'un-', prefixedOnly: true }).extract({
        content: fixtureOnlyPrefix,
        cursor: 15,
      })

      expect(res1).toBeNull()

      const res2 = await autocompleteExtractorAttributify({ prefix: 'un-', prefixedOnly: true }).extract({
        content: fixtureOnlyPrefix,
        cursor: 34,
      })

      expect(res2).toBeNull()
    })
  })

  it('with trueToNonValued', async () => {
    const uno = await createGenerator({
      presets: [
        presetAttributify({ trueToNonValued: true }),
        presetWind3(),
      ],
    })
    const { css } = await uno.generate(`
      <div flex="true"></div>
      <div grid="~ cols-2"></div>
    `, { preflights: false })
    expect(css).toMatchSnapshot()
  })

  it('with incomplete element', async () => {
    await uno.generate('<div class="w-fullllllllllllll"')
  }, 20)

  it('merge attribute name and value-only', async () => {
    const { css } = await uno.generate(`
      <div bg="[&:nth-child(3)]:[#123456]"></div>
      <div class="foo" bg="[&.foo]:[&:nth-child(3)]:[#123]"></div>
    `, { preflights: false })
    expect(css).toMatchSnapshot()
  })

  it('support inline arrow functions', async () => {
    const uno = await createGenerator({
      presets: [
        presetAttributify(),
        presetWind3(),
      ],
    })
    const { css: css1 } = await uno.generate('<div v-for="(v, i) of [0].map(() => 1)" h-1px />', { preflights: false })
    expect(css1).toMatchSnapshot()

    const { css: css2 } = await uno.generate(`
      <div
        h-1px
        v-for="(
          v, i
        ) of [0].map(() => 1)"
      />
    `, { preflights: false })
    expect(css2).toMatchSnapshot()

    const { css: css3 } = await uno.generate(`
      <div v-for="(
          v, i
        ) of [0].map(() => 1)"
        h-1px
      />
    `, { preflights: false })
    expect(css3).toMatchSnapshot()
  })
})
