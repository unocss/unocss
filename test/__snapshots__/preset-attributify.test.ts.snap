// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`attributify > merge attribute name and value-only 1`] = `
"/* layer: default */
[bg~="\\[\\&\\:nth-child\\(3\\)\\]\\:\\[\\#123456\\]"]:nth-child(3){--un-bg-opacity:1;background-color:rgb(18 52 86 / var(--un-bg-opacity));}
[bg~="\\[\\&\\.foo\\]\\:\\[\\&\\:nth-child\\(3\\)\\]\\:\\[\\#123\\]"].foo:nth-child(3){--un-bg-opacity:1;background-color:rgb(17 34 51 / var(--un-bg-opacity));}"
`;

exports[`attributify > prefixedOnly 1`] = `
"/* layer: default */
.absolute{position:absolute;}
.fixed{position:fixed;}
[un-children~="m-auto"]>*{margin:auto;}"
`;

exports[`attributify > support inline arrow functions 1`] = `
"/* layer: default */
.h-1px,
[h-1px=""]{height:1px;}"
`;

exports[`attributify > support inline arrow functions 2`] = `
"/* layer: default */
.h-1px,
[h-1px=""]{height:1px;}"
`;

exports[`attributify > support inline arrow functions 3`] = `
"/* layer: default */
.h-1px,
[h-1px=""]{height:1px;}"
`;

exports[`attributify > variant 1`] = `
[
  "uno-layer-base-c-white/10",
  "hover:uno-layer-base-c-black/20",
  "sm-[color:red]",
  "md-[--var:var(--another)]",
  "lg-bg-blue-600",
  undefined,
  undefined,
  "important-text-red",
  "important-bg-red",
  "bg-blue-400",
  "hover:bg-blue-500",
  "dark:!bg-blue-500",
  "dark:hover:bg-blue-600",
  "text-sm",
  "text-white",
  "!leading-4",
  "!flex",
  "flex-col",
  "flex-items-center",
  "flex-justify-center",
  "p-t-2",
  "pt-2",
  "border-rounded-xl",
  "border-x-1",
  "border-x-style-dashed",
  "font-mono",
  "font-sans",
  "p-y-2",
  "p-x-4",
  "border-2",
  "border-rounded",
  "border-blue-200",
  "children-m-auto",
  "x-transition:enter-start-opacity-0",
  "pt2",
  "rounded-sm",
  "size-w-1",
  "size-h-0.5",
  "inline-block",
  "transform",
  "translate-x-100%",
  "translate-y-",
  "%",
  "rotate-30",
  "after-content-[quoted:uno_css]",
  "rotate-60",
  "ma",
  "m-\\",
  "m-1",
  "m-2",
  "m-3",
  "ml-1.5",
  "ml-1..6",
]
`;

exports[`attributify > with trueToNonValued 1`] = `
"/* layer: default */
[grid~="\\~"]{display:grid;}
.cols-2,
[grid~="cols-2"]{grid-template-columns:repeat(2,minmax(0,1fr));}
[flex~="true"]{display:flex;}"
`;
